File: python/luhn/luhn.py
import re

class Luhn(object):
	def __init__(self, card_num: str):
		self.card_num = re.sub(' ', '', card_num.lower())

	def valid(self) -> bool:
		if len(self.card_num) <= 1: 
			return False

		if not self.has_only_digit_chars(self.card_num): 
			return False

		return self.calculate_luhn(self.card_num) == 0

	def has_only_digit_chars(self, card_num: str) -> bool:
		return len(re.sub('[0-9]', '', card_num)) == 0

	def calculate_luhn(self, card_num: str) -> int:
		return sum(list(map(lambda tup: int(tup[1]) if tup[0] % 2 == 0 else self.double(int(tup[1])), enumerate(card_num[::-1])))) % 10

	def double(self, num: int) -> int:
		return num * 2 if num * 2 < 10 else (num * 2) - 9

File: go/luhn/luhn.go
// Package luhn contains a method validating a number matches the luhn algorithm.
package luhn

import (
	"strings"
	"unicode"
)

// Valid validates the input number is a valid luhn number.
func Valid(input string) bool {
	input = removeSpaces(input)

	if len(input) < 2 {
		return false
	}

	return validLuhnNumber(input)
}

func removeSpaces(input string) string {
	return strings.ReplaceAll(input, " ", "")
}

func validLuhnNumber(number string) bool {
	sum := 0

	for i, digit := range number {
		if !unicode.IsDigit(digit) {
			return false
		}

		num := int(digit - '0')
		if shouldDouble(i, len(number)) {
			sum += double(num)
		} else {
			sum += num
		}
	}

	return sum%10 == 0
}

func shouldDouble(currIndex, totalLength int) bool {
	if totalLength%2 == 1 && currIndex%2 == 1 {
		return true
	}

	if totalLength%2 == 0 && currIndex%2 == 0 {
		return true
	}

	return false
}

func double(digit int) (doubled int) {
	doubled = digit * 2

	if doubled > 9 {
		doubled -= 9
	}

	return
}
